### ISSUE TITLE
macbook thunderbolt issue

### ISSUE THREAD

**Issue by mfauvain** [intel/thunderbolt-software-user-space#65]
Hi, having an issue with macbookpro14,1 running linux (kernel 4.16) and thunderbolt display (connected via TB3 to 2 apple adapter), if not right forum let me know. Was using a thunderbolt display with some connected devices (usb and firewire) and everything was working well (facetime camera, sound, ethernet, and all connected devices) up until a week ago when only display was recognized. Nothing was changed on the laptop, no update, nada. I've been searching a lot but no answer so far. Only display is recognized (and working, even 2 thunderbolt display daisy chained are working fine, except with no devices, no facetime, no ethernet, no sound). 
only clue in dmesg  is `[   27.895901] thunderbolt 0000:06:00.0: 4:c: disabled by eeprom` etc..

tbtadm shows tbdisplay is authorized. tried bolt as well and tbdisplay is authorized. tried reset SMC, NVRAM no luck.
Any clue?

EDIT : for completness here is the tbtadm topology output : 
`Controller 0`
`    ├─ Details:`
`    │   ├─ Name: Macintosh, Apple Inc.`
`    │   └─ Security level: SL0 (none)`
`    │`
`    └─ Thunderbolt Display, Apple, Inc.`
`        └─ Details:`
`            ├─ Route-string: 0-4`
`            ├─ Authorized: Yes`
`            ├─ In ACL: No`
`            └─ UUID: 90ad5500-0100-0100-ffff-ffffffffffff`

EDIT2 : also, booting OSX : all devices (ethernet, sound, facetime, firewire and USB) and displays shown and working

**Comment by westeri** [comment]
So on Apple system you don't need tbtadm or bolt because the firmware connects all devices automatically.

You are saying that display port works but PCIe does not? There might be some issue with the PCIe hotplug. Can you share your full dmesg when you have plugged in a device and output of 'sudo lspci -vv'?

**Comment by mfauvain** [comment]
correct. thunderbolt display (connected to tb3 to tb2 adapter) works, even 2 daisy chained display works, however none of the devices connected (or integrated) to the displays are recognized anymore. Here are the lspci -vv and dmesg
[dmesg.txt](https://github.com/intel/thunderbolt-software-user-space/files/2657317/dmesg.txt)
[lspci.txt](https://github.com/intel/thunderbolt-software-user-space/files/2657318/lspci.txt)

**Comment by westeri** [comment]
Can you remove `pci=hpbussize=15,hpmemsize=10M` from the kernel command line (you should never need to pass any custom command line options to the kernel) and then retry so that you boot the system without the device connected, then plug it in once the system is up and running? Please provide dmesg from that and `sudo lspci -vv` (note `sudo`).

**Comment by mfauvain** [comment]
sure, here are the files
[dmesg.txt](https://github.com/intel/thunderbolt-software-user-space/files/2657387/dmesg.txt)
[sudolspci.txt](https://github.com/intel/thunderbolt-software-user-space/files/2657388/sudolspci.txt)

**Comment by westeri** [comment]
Hmm, I don't see any errors in the dmesg that could be related. The device itself is detected just fine but it looks like PCIe is not tunneled for some reason. If you check `/sys/bus/thunderbolt/devices/0-4/authorized` does it contain `1`?

**Comment by mfauvain** [comment]
indeed, `/sys/bus/thunderbolt/devices/0-4/authorized` is `1`
in dmesg, lines saying disabled by eeprom are fine?

**Comment by westeri** [comment]
Yes, those are fine.

**Comment by westeri** [comment]
When it stopped working, are you sure nothing was changed like for example kernel got upgraded by the distro?

**Comment by mfauvain** [comment]
positive. I use an arch repo from archive (to have kernel 4.16, later kernels dont boot on this macbook)  and everything is fixed. what I don't get is, usb key used to install arch on this machine now present the same problem. When I initialy booted form this usb key, thunderbolt display and all devices were present and working, after install (booting the installed arch) tbdisplay and devices connected to it were working. Now even when booting the USB key, tbdisplay is working but none of the connected devices are. Was something written to mac bios? if it was, how come booting OSX get the tbdisplay AND connected devices working? is bios detecting when non-mac os is running and change security level of thunderbolt?

**Comment by westeri** [comment]
macOS uses different approach to connected devices. It creates tunnels via software but when you boot Linux we start the firmware which does the tunnel creation. To me it looks like the firmware got the tunnels up and running but something prevents the PCIe hotplug event from propagating. One thing to try is to run manually (as root) `echo 1 > /sys/bus/pci/rescan`. That rescans the whole PCI bus. If it finds the devices then at least the PCIe tunnel is up.

**Comment by mfauvain** [comment]
I did try this before, and just did again for mental sanity and no luck. Just to make sure, `sudo echo 1 > /sys/bus/pci/rescan` gives permission denied so I did `sudo -i` and then `echo 1 > /sys/bus/pci/rescan` which doen't resturn anything right?

**Comment by westeri** [comment]
Right, it does not return anything but you should see something new in the dmesg and `lspci` output if it finds anything.

**Comment by mfauvain** [comment]
was actually looking at dmesg and have this (right after echo 1 > ...)
`[ 3781.086186] pcieport 0000:05:02.0: bridge window [mem 0x00100000-0x000fffff 64bit pref] to [bus 07] add_size 200000 add_align 100000`
`[ 3781.086241] pcieport 0000:05:02.0: BAR 15: no space for [mem size 0x00200000 64bit pref]`
`[ 3781.086243] pcieport 0000:05:02.0: BAR 15: failed to assign [mem size 0x00200000 64bit pref]`
`[ 3781.086247] pcieport 0000:05:02.0: BAR 15: no space for [mem size 0x00200000 64bit pref]`
`[ 3781.086248] pcieport 0000:05:02.0: BAR 15: failed to assign [mem size 0x00200000 64bit pref]`

**Comment by mfauvain** [comment]
not familiar at all with this but, the kernel params I had set after the issue appeared the first time was `pci=hpbussize=10,hpmemsize=2M` and looking at the dmesg I am wondering if the part `add_size 200000 add_align 100000` correspond to this? is kernel param stuck somewhere?

**Comment by westeri** [comment]
It is the default the kernel uses for MMIO if nothing is specified. Those messages in dmesg are fine.

However, for some reason the PCIe tunnel is not established properly and there is no indication in the log what might be wrong :(

**Comment by mfauvain** [comment]
thanks a lot for the help. I tried booting another linux distro and same issue. very frustrating.

**Comment by mfauvain** [comment]
As another check, I plugged my previous laptop macbookpro 10,2 to the same thunderbolt display (this 10,2 has tb2 port, so no adapter TB3 to TB2 this time) and everything works (ie devices connected to TBdisplay are recognized and working)

**Comment by westeri** [comment]
With TB2 in Linux you can connect one device but there is no Display Port tunneling so even if PCIe devices of the display work there is no picture.

Can you do one test on the TB3 macbook? In the Thunderbolt driver, there is a function that gets called whenever TBT device is connected:
```
drivers/thunderbolt/icm.c::icm_fr_device_connected()
{
   ...
   authorized = pkg->link_info & ICM_LINK_INFO_APPROVED;
```
If you comment out that line (`//`), connect a device and then do
```
   sudo sh -c 'echo 1 > /sys/bus/thunderbolt/devices/0-4/authorized'
```
Does it work? I kind of assume it does not but it is also possible that the firmware is reporting the authorization status wrong.

**Comment by mfauvain** [comment]
sorry, where should the driver be? /sys/bus/thunderbolt/drivers is empty.

**Comment by westeri** [comment]
I mean the driver in Linux kernel source code (lives in `drivers/thunderbolt` directory). Are you familiar how to build your own kernel? If not, then I suppose it is not worth a try.

**Comment by mfauvain** [comment]
I am not, but will try and revert. I initially posted the issue on another forum [https://github.com/Dunedan/mbp-2016-linux/issues/87](url) and I am following on there as well, but would ask the same question here : 
Question that bothers me though is : on day one I used a usb key to install arch on this machine and on that day, thunderbolt display + all connected devices were working (actually the usb key was plugged into thunderbolt display usb port) I even used the external keyboard plugged to that same display to do all the install as internal keyboard was not recognized before installing macbook12-spi-driver. How come today booting the same usb key (no changes at all on that key) has only display working, not the keyboard, not even the usb port : holding option key at startup shows the key, booting starts and then fails right after kernel tries to read the usb key to follow on booting process.. would the tb3 to tb2 adapter be blocking this now??? can anything be 'written' to that adpter's chip?

**Comment by westeri** [comment]
I don't think the adapter has anything to do with the issue. I've been using Apple TB3<->TB2 adapter and in general it works just fine.

**Comment by YehezkelShB** [comment]
I think some of the non-Apple TB3<->TB2 adapters don't work with Apple TBT host controllers, due to PD FW incompatibility

**Comment by mfauvain** [comment]
Compiled kernel with the line ` authorized = pkg->link_info & ICM_LINK_INFO_APPROVED;` commented out. and running command `sudo sh -c 'echo 1 > /sys/bus/thunderbolt/devices/0-4/authorized'` returns : 
`sh: line 0: echo: write error: Input/output error`.

boltclt shows 
`● Apple Thunderbolt Display`
`   ├─ type:          peripheral`
`   ├─ name:          Thunderbolt Display`
`   ├─ vendor:        Apple, Inc.`
`   ├─ uuid:          90ad5500-0100-0100-ffff-ffffffffffff`
`   ├─ status:        authorization error`
`   │  ├─ domain:     d0010000-0080-7f08-235f-4312aa32a021`
`   │  └─ authflags:  none`
`   ├─ authorized:    Sat 08 Dec 2018 23:38:56 UTC`
`   ├─ connected:     Sat 08 Dec 2018 23:38:56 UTC`
`   └─ stored:        no`

and dmesg 

`[   26.471878] thunderbolt 0000:06:00.0: current switch config:`
`[   26.471888] thunderbolt 0000:06:00.0:  Switch: 8086:1513 (Revision: 2, TB Version: 1)`
`[   26.471892] thunderbolt 0000:06:00.0:   Max Port Number: 13`
`[   26.471895] thunderbolt 0000:06:00.0:   Config:`
`[   26.471901] thunderbolt 0000:06:00.0:    Upstream Port Number: 2 Depth: 1 Route String: 0x4 Enabled: 1, PlugEventsDelay: 254ms`
`[   26.471905] thunderbolt 0000:06:00.0:    unknown1: 0x0 unknown4: 0x0`
`[   26.509374] thunderbolt 0000:06:00.0: 4: reading drom (length: 0x97)`
`[   27.018264] thunderbolt 0000:06:00.0: 4: uid: 0x100010055ad90`
`[   27.018390] thunderbolt 0000:06:00.0:  Port 0: 8086:1513 (Revision: 2, TB Version: 1, Type: Port (0x1))`
`[   27.018392] thunderbolt 0000:06:00.0:   Max hop id (in/out): 7/7`
`[   27.018394] thunderbolt 0000:06:00.0:   Max counters: 8`
`[   27.018396] thunderbolt 0000:06:00.0:   NFC Credits: 0x700000`
`[   27.019029] thunderbolt 0000:06:00.0:  Port 1: 8086:1513 (Revision: 2, TB Version: 1, Type: Port (0x1))`
`[   27.019031] thunderbolt 0000:06:00.0:   Max hop id (in/out): 63/63`
`[   27.019033] thunderbolt 0000:06:00.0:   Max counters: 32`
`[   27.019035] thunderbolt 0000:06:00.0:   NFC Credits: 0x3c0001c`
`[   27.019668] thunderbolt 0000:06:00.0:  Port 2: 8086:1513 (Revision: 2, TB Version: 1, Type: Port (0x1))`
`[   27.019670] thunderbolt 0000:06:00.0:   Max hop id (in/out): 63/63`
`[   27.019672] thunderbolt 0000:06:00.0:   Max counters: 32`
`[   27.019674] thunderbolt 0000:06:00.0:   NFC Credits: 0x3c00000`
`[   27.020309] thunderbolt 0000:06:00.0:  Port 3: 8086:1513 (Revision: 2, TB Version: 1, Type: Port (0x1))`
`[   27.020311] thunderbolt 0000:06:00.0:   Max hop id (in/out): 63/63`
`[   27.020313] thunderbolt 0000:06:00.0:   Max counters: 32`
`[   27.020315] thunderbolt 0000:06:00.0:   NFC Credits: 0x3c00000`
`[   27.020948] thunderbolt 0000:06:00.0:  Port 4: 8086:1513 (Revision: 2, TB Version: 1, Type: Port (0x1))`
`[   27.020950] thunderbolt 0000:06:00.0:   Max hop id (in/out): 63/63`
`[   27.020952] thunderbolt 0000:06:00.0:   Max counters: 32`
`[   27.020954] thunderbolt 0000:06:00.0:   NFC Credits: 0x3c00000`
`[   27.020957] thunderbolt 0000:06:00.0: 4:5: disabled by eeprom`
`[   27.020959] thunderbolt 0000:06:00.0: 4:6: disabled by eeprom`
`[   27.021077] thunderbolt 0000:06:00.0:  Port 7: 8086:1513 (Revision: 2, TB Version: 1, Type: PCIe (0x100101))`
`[   27.021079] thunderbolt 0000:06:00.0:   Max hop id (in/out): 8/8`
`[   27.021081] thunderbolt 0000:06:00.0:   Max counters: 1`
`[   27.021083] thunderbolt 0000:06:00.0:   NFC Credits: 0x700000`
`[   27.027856] thunderbolt 0000:06:00.0:  Port 8: 8086:1513 (Revision: 2, TB Version: 1, Type: PCIe (0x100101))`
`[   27.027859] thunderbolt 0000:06:00.0:   Max hop id (in/out): 8/8`
`[   27.027861] thunderbolt 0000:06:00.0:   Max counters: 1`
`[   27.027863] thunderbolt 0000:06:00.0:   NFC Credits: 0x700000`
`[   27.027865] thunderbolt 0000:06:00.0: 4:9: disabled by eeprom`
`[   27.027980] thunderbolt 0000:06:00.0:  Port 10: 8086:1513 (Revision: 2, TB Version: 1, Type: PCIe (0x100102))`
`[   27.027982] thunderbolt 0000:06:00.0:   Max hop id (in/out): 8/8`
`[   27.027984] thunderbolt 0000:06:00.0:   Max counters: 1`
`[   27.027986] thunderbolt 0000:06:00.0:   NFC Credits: 0x700000`
`[   27.028157] thunderbolt 0000:06:00.0:  Port 11: 8086:1513 (Revision: 2, TB Version: 1, Type: DP/HDMI (0xe0102))`
`[   27.028160] thunderbolt 0000:06:00.0:   Max hop id (in/out): 9/9`
`[   27.028162] thunderbolt 0000:06:00.0:   Max counters: 2`
`[   27.028165] thunderbolt 0000:06:00.0:   NFC Credits: 0x700000`
`[   27.028167] thunderbolt 0000:06:00.0: 4:c: disabled by eeprom`
`[   27.028169] thunderbolt 0000:06:00.0: 4:d: disabled by eeprom`
`[   71.237616] thunderbolt 0000:06:00.0: PCIe tunnel creation failed`

display still works but connected devices don't.

**Comment by westeri** [comment]
Yeah, I kind of expected that to happen :( OK, so the firmware says it has created the PCIe tunnel but for unknown reason there is no PCIe link.

**Comment by mfauvain** [comment]
thx @westeri, as display works but other devices connected to this apple tbt display don't, I understand the port is stuck in dponly mode and there is no way (software or NVRAM reset etc..) to set it back to security mode 'none'. Am I correct?

**Comment by westeri** [comment]
Well it is already on `none` mode. What you can try is to use the SW CM implementation instead but then you lose DP tunneling. In order to do so, you need to make `drivers/thunderbolt/icm.c::icm_ar_is_supported()` return false:
```
static bool icm_ar_is_supported(struct tb *tb)
{
        return false; // HACK HACK

        /*
         * Starting from Alpine Ridge we can use ICM on Apple machines
         * as well. We just need to reset and re-enable it first.
         */
         ...
```

**Comment by mfauvain** [comment]
I will try, however I prefer to have only display than only devices..
